File: DNA Cryptography – Inspired by Biology.go
package main

import (
	"fmt"
	"math/rand"
	"strings"
	"time"
)

// DNA Mapping Tables
var binaryToDNA = map[string]rune{
	"00": 'A',
	"01": 'T',
	"10": 'C',
	"11": 'G',
}

var dnaToBinary = map[rune]string{
	'A': "00",
	'T': "01",
	'C': "10",
	'G': "11",
}

// Convert Text to Binary
func textToBinary(text string) string {
	binaryString := ""
	for _, c := range text {
		binaryString += fmt.Sprintf("%08b", c)
	}
	return binaryString
}

// Convert Binary to DNA
func binaryToDna(binary string) string {
	var dna strings.Builder
	for i := 0; i < len(binary); i += 2 {
		dna.WriteRune(binaryToDNA[binary[i:i+2]])
	}
	return dna.String()
}

// Convert DNA to Binary
func dnaToBinaryString(dna string) string {
	var binary strings.Builder
	for _, nucleotide := range dna {
		binary.WriteString(dnaToBinary[nucleotide])
	}
	return binary.String()
}

// Generate Random DNA Key
func generateDnaKey(length int) string {
	nucleotides := "ATCG"
	rand.Seed(time.Now().UnixNano())
	var key strings.Builder
	for i := 0; i < length; i++ {
		key.WriteByte(nucleotides[rand.Intn(4)])
	}
	return key.String()
}

// XOR-Based DNA Encryption
func dnaXorEncrypt(dna, key string) string {
	var encrypted strings.Builder
	dnaMap := "ATCG"
	for i := range dna {
		dnaBin := strings.IndexRune(dnaMap, rune(dna[i]))
		keyBin := strings.IndexRune(dnaMap, rune(key[i]))
		encrypted.WriteByte(dnaMap[dnaBin^keyBin]) // XOR operation
	}
	return encrypted.String()
}

// XOR-Based DNA Decryption
func dnaXorDecrypt(encryptedDna, key string) string {
	var decrypted strings.Builder
	dnaMap := "ATCG"
	for i := range encryptedDna {
		encBin := strings.IndexRune(dnaMap, rune(encryptedDna[i]))
		keyBin := strings.IndexRune(dnaMap, rune(key[i]))
		decrypted.WriteByte(dnaMap[encBin^keyBin]) // Reverse XOR
	}
	return decrypted.String()
}

// Convert Binary to Text
func binaryToText(binary string) string {
	var text strings.Builder
	for i := 0; i < len(binary); i += 8 {
		charValue := 0
		fmt.Sscanf(binary[i:i+8], "%b", &charValue)
		text.WriteByte(byte(charValue))
	}
	return text.String()
}

// Encrypt Message
func encryptMessage(message string) (string, string) {
	binaryData := textToBinary(message)
	dnaSequence := binaryToDna(binaryData)
	key := generateDnaKey(len(dnaSequence))
	encryptedDna := dnaXorEncrypt(dnaSequence, key)
	return encryptedDna, key
}

// Decrypt Message
func decryptMessage(encryptedDna, key string) string {
	decryptedDna := dnaXorDecrypt(encryptedDna, key)
	binaryData := dnaToBinaryString(decryptedDna)
	return binaryToText(binaryData)
}

// Main Function
func main() {
	message := "HELLO DNA"
	fmt.Println("Original Message:", message)

	// Encrypt
	encryptedDna, dnaKey := encryptMessage(message)
	fmt.Println("Encrypted DNA:", encryptedDna)
	fmt.Println("DNA Key:      ", dnaKey)

	// Decrypt
	decryptedMessage := decryptMessage(encryptedDna, dnaKey)
	fmt.Println("Decrypted Message:", decryptedMessage)
}

File: DNA Cryptography – Inspired by Biology.py
import random

# DNA Mapping for Encoding and Decoding
binary_to_dna = {
    "00": "A",
    "01": "T",
    "10": "C",
    "11": "G"
}

dna_to_binary = {v: k for k, v in binary_to_dna.items()}

# Generate a Random DNA Key
def generate_dna_key(length):
    return ''.join(random.choice("ATCG") for _ in range(length))

# Convert Text to Binary
def text_to_binary(text):
    return ''.join(format(ord(char), '08b') for char in text)

# Convert Binary to DNA
def binary_to_dna_sequence(binary):
    return ''.join(binary_to_dna[binary[i:i+2]] for i in range(0, len(binary), 2))

# Convert DNA to Binary
def dna_to_binary_sequence(dna):
    return ''.join(dna_to_binary[nucleotide] for nucleotide in dna)

# XOR-Based DNA Encryption
def dna_xor_encrypt(dna_sequence, dna_key):
    encrypted_dna = []
    for i in range(len(dna_sequence)):
        encrypted_dna.append(binary_to_dna[str(int(dna_to_binary[dna_sequence[i]]) ^ int(dna_to_binary[dna_key[i]]))])
    return ''.join(encrypted_dna)

# XOR-Based DNA Decryption
def dna_xor_decrypt(encrypted_dna, dna_key):
    decrypted_dna = []
    for i in range(len(encrypted_dna)):
        decrypted_dna.append(binary_to_dna[str(int(dna_to_binary[encrypted_dna[i]]) ^ int(dna_to_binary[dna_key[i]]))])
    return ''.join(decrypted_dna)

# Convert Binary to Text
def binary_to_text(binary):
    return ''.join(chr(int(binary[i:i+8], 2)) for i in range(0, len(binary), 8))

# Encrypt a Message Using DNA Cryptography
def encrypt_message(message):
    binary_data = text_to_binary(message)
    dna_sequence = binary_to_dna_sequence(binary_data)
    dna_key = generate_dna_key(len(dna_sequence))
    encrypted_dna = dna_xor_encrypt(dna_sequence, dna_key)
    return encrypted_dna, dna_key

# Decrypt a Message Using DNA Cryptography
def decrypt_message(encrypted_dna, dna_key):
    decrypted_dna = dna_xor_decrypt(encrypted_dna, dna_key)
    binary_data = dna_to_binary_sequence(decrypted_dna)
    return binary_to_text(binary_data)

# Test the DNA Cryptography System
if __name__ == "__main__":
    message = "HELLO DNA"
    print(f"Original Message: {message}")

    # Encrypt
    encrypted_dna, dna_key = encrypt_message(message)
    print(f"Encrypted DNA: {encrypted_dna}")
    print(f"DNA Key:       {dna_key}")

    # Decrypt
    decrypted_message = decrypt_message(encrypted_dna, dna_key)
    print(f"Decrypted Message: {decrypted_message}")
